### DISCUSSION TITLE
How to filter out historical entities

### DISCUSSION THREAD

**Discussion by szeke** [usc-isi-i2/kgtk#472] [Q&A] [answered]
Background: Wikidata has many historical countries and admin regions. They have a `P31` to a subclass of `Q19832712 `. I would like to filter out all historical admins. 

Given an input graph G I want to create an output graph J where entities that satisfy the following are removed.
- (x)-[:P31]->(class)
- (class)-[:P279star]->(: Q19832712)

But, actually, the above does not work because the USA (Q30) is an instance of historical country, but the P31 statement has an end date. So, I really want to remove:

- (x)-[r:P31]->(class)
- (class)-[:P279star]->(: Q19832712)
- r has no end time (P582) qualifier

**Comment by chalypso** [comment, marked as answer]
This is a not-exists query which (for now) can be handled with an optional match pattern like so (the --where: is a global where which is evaluated outside the scope of any of the match clauses which is why we can check for NULL there):
```
> kgtk query --graph-cache wikidata-20210215-dwd-browser.2021-07-09T19:19.sqlite3.db \
          -i claims -i p279star -i qualifiers\
          --match 'claims: (x)-[r:P31]->(class), p279: (class)-[]->(:Q19832712)' \
          --opt   'qual: (r)-[:P582]->(endtime)' \
          --where: 'endtime is NULL' \
          --return 'x, class' \
          --limit 10
node1	node2
Q17832526	Q10007123
Q2547918	Q10007123
Q5157575	Q10007123
Q13111272	Q101085149
Q13113437	Q101085149
Q16133687	Q101085149
Q16133855	Q101085149
Q20580585	Q101085149
Q23307849	Q101085149
Q98546017	Q101085149
```
